### PR TITLE
store/tikv,executor: redesign the latch scheduler

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

The old latch scheduler checks `maxCommitTS` on each slot, different key could hash to the same slot,
then the latch scheduler will report a fake transaction confliction.

### What is changed and how it works?

Each key will have their own `maxCommitTS`, so hash collision will not result in a fake transaction confliction.

### Check List 

Tests 

 - Unit test

Code changes

- remove `RefreshCommitTS`
- redesign latch structure

PTAL @coocood

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @coocood @disksing

File: store/tikv/latch/latch.go [hunk 1]
```diff
@@ -161,75 +156,99 @@ func (latches *Latches) acquire(lock *Lock) acquireResult {
...
 	defer latch.Unlock()
-	if latch.maxCommitTS > lock.startTS {
+
+	var find *node
+	for n := latch.queue; n != nil; n = n.next {
+		if bytes.Compare(n.key, key) == 0 {
+			find = n
+			break
+		}
+		// TODO: Invalidate old data.
```
Comment (disksing) [inline store/tikv/latch/latch.go]: We cannot merge this PR before fixing this TODO. Because without it we will get an OOM.

**Comment by tiancaiamao** [comment]
PTAL @disksing @coocood

**Comment by tiancaiamao** [comment]
PTAL @disksing @coocood

File: store/tikv/latch/latch.go [hunk 2]
```diff
@@ -22,32 +23,26 @@ import (
...
-}
+type node struct {
+	slotID      int
+	key         []byte
+	maxCommitTS uint64
+	value       *Lock
 
-func (l *latch) isEmpty() bool {
-	return l.waitingQueueHead == 0 && !l.hasMoreWaiting
+	next *node
```
Comment (disksing) [inline store/tikv/latch/latch.go]: Why not use `list.List`?

File: store/tikv/latch/latch.go [hunk 3]
```diff
@@ -161,75 +157,136 @@ func (latches *Latches) acquire(lock *Lock) acquireResult {
...
+		latch.queue = tmp
+		latch.count++
+
+		lock.acquiredCount++
+		return acquireSuccess
+	}
+
+	// Try to limits the memory usage.
+	if latch.count > latchListCount {
+		latch.recycle(lock.startTS)
```
Comment (disksing) [inline store/tikv/latch/latch.go]: Should `recycle`be moved into `if find == nil {}`?

File: store/tikv/latch/latch.go [hunk 4]
```diff
@@ -161,75 +157,136 @@ func (latches *Latches) acquire(lock *Lock) acquireResult {
...
-	latch.waitingQueueHead = nextLock.startTS
-	nextLock.acquiredCount++
-	if latch.maxCommitTS > nextLock.startTS {
-		nextLock.isStale = true
+
+	idx := 0
+	for i := 0; i < len(latch.waiting); i++ {
+		waiting := latch.waiting[i]
+		if bytes.Compare(waiting.keys[waiting.acquiredCount], key) == 0 {
+			nextLock = waiting
```
Comment (disksing) [inline store/tikv/latch/latch.go]: Is it possible that there are more than 1 `Lock`s in the waiting list have the same key?

**Comment by disksing** [comment]
Do we have any test/benchmark result?

**Comment by disksing** [comment]
PTAL @zhangjinpeng1987

**Comment by tiancaiamao** [comment]
I use a `update_non_index` test case from sysbench, the table schema:

```
CREATE TABLE `sbtest1` (
  `id` int(10) UNSIGNED NOT NULL AUTO_INCREMENT,
  `k` int(10) UNSIGNED NOT NULL DEFAULT '0',
  `c` char(120) NOT NULL DEFAULT '',
  `pad` char(60) NOT NULL DEFAULT '',
  PRIMARY KEY (`id`),
  KEY `k_1` (`k`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin AUTO_INCREMENT=30001
```

key range of `k` is 16, test concurrency with 256, so there will be a lot of conflicts during the update operation.

When latch scheduler is disabled:

```
sysbench --test=./lua-tests/db/update_non_index.lua --db-driver=mysql --mysql-host=127.0.0.1 --mysql-port=4000 --mysql-user=root --mysql-password= --mysql-db=sbtest1 --oltp-tables-count=1 --oltp-table-size=16 --num-threads=256 --report-interval=10 --max-requests=2000000000 --percentile=95 --max-time=300 run
WARNING: the --test option is deprecated. You can pass a script name or path on the command line without any options.
WARNING: --num-threads is deprecated, use --threads instead
WARNING: --max-requests is deprecated, use --events instead
WARNING: --max-time is deprecated, use --time instead
sysbench 1.0.11 (using system LuaJIT 2.1.0-beta3)

Running the test with following options:
Number of threads: 256
Report intermediate results every 10 second(s)
Initializing random number generator from current time


Initializing worker threads...

Threads started!

[ 10s ] thds: 256 tps: 6.10 qps: 6.10 (r/w/o: 0.00/6.10/0.00) lat (ms,95%): 8638.96 err/s: 0.00 reconn/s: 0.00
[ 20s ] thds: 256 tps: 2.60 qps: 2.60 (r/w/o: 0.00/2.60/0.00) lat (ms,95%): 19078.64 err/s: 0.00 reconn/s: 0.00
[ 30s ] thds: 256 tps: 3.10 qps: 3.10 (r/w/o: 0.00/3.10/0.00) lat (ms,95%): 27846.48 err/s: 0.00 reconn/s: 0.00
[ 40s ] thds: 256 tps: 3.60 qps: 3.60 (r/w/o: 0.00/3.60/0.00) lat (ms,95%): 38506.38 err/s: 0.00 reconn/s: 0.00
[ 50s ] thds: 256 tps: 5.00 qps: 5.00 (r/w/o: 0.00/5.00/0.00) lat (ms,95%): 49546.69 err/s: 0.00 reconn/s: 0.00
[ 60s ] thds: 256 tps: 10.90 qps: 10.90 (r/w/o: 0.00/10.90/0.00) lat (ms,95%): 56202.50 err/s: 0.00 reconn/s: 0.00
[ 70s ] thds: 256 tps: 1.20 qps: 1.20 (r/w/o: 0.00/1.20/0.00) lat (ms,95%): 66090.17 err/s: 0.00 reconn/s: 0.00
[ 80s ] thds: 256 tps: 1.50 qps: 1.50 (r/w/o: 0.00/1.50/0.00) lat (ms,95%): 76330.47 err/s: 0.00 reconn/s: 0.00
[ 90s ] thds: 256 tps: 1.80 qps: 1.80 (r/w/o: 0.00/1.80/0.00) lat (ms,95%): 82031.09 err/s: 0.00 reconn/s: 0.00
[ 100s ] thds: 256 tps: 1.70 qps: 1.70 (r/w/o: 0.00/1.70/0.00) lat (ms,95%): 96462.77 err/s: 0.00 reconn/s: 0.00
[ 110s ] thds: 256 tps: 1.40 qps: 1.40 (r/w/o: 0.00/1.40/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 120s ] thds: 256 tps: 1.70 qps: 1.70 (r/w/o: 0.00/1.70/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 130s ] thds: 256 tps: 1.90 qps: 1.90 (r/w/o: 0.00/1.90/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 140s ] thds: 256 tps: 2.60 qps: 2.60 (r/w/o: 0.00/2.60/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 150s ] thds: 256 tps: 2.10 qps: 2.10 (r/w/o: 0.00/2.10/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 160s ] thds: 256 tps: 1.50 qps: 1.50 (r/w/o: 0.00/1.50/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 170s ] thds: 256 tps: 1.50 qps: 1.50 (r/w/o: 0.00/1.50/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 180s ] thds: 256 tps: 1.30 qps: 1.30 (r/w/o: 0.00/1.30/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 190s ] thds: 256 tps: 1.20 qps: 1.20 (r/w/o: 0.00/1.20/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 200s ] thds: 256 tps: 1.60 qps: 1.60 (r/w/o: 0.00/1.60/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 210s ] thds: 256 tps: 1.80 qps: 1.80 (r/w/o: 0.00/1.80/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 220s ] thds: 256 tps: 1.80 qps: 1.80 (r/w/o: 0.00/1.80/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 230s ] thds: 256 tps: 1.90 qps: 1.90 (r/w/o: 0.00/1.90/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 240s ] thds: 256 tps: 1.90 qps: 1.90 (r/w/o: 0.00/1.90/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 250s ] thds: 256 tps: 1.10 qps: 1.10 (r/w/o: 0.00/1.10/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 260s ] thds: 256 tps: 1.70 qps: 1.70 (r/w/o: 0.00/1.70/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 270s ] thds: 256 tps: 1.90 qps: 1.90 (r/w/o: 0.00/1.90/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 280s ] thds: 256 tps: 1.70 qps: 1.70 (r/w/o: 0.00/1.70/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 290s ] thds: 256 tps: 1.40 qps: 1.40 (r/w/o: 0.00/1.40/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 300s ] thds: 256 tps: 3.30 qps: 3.30 (r/w/o: 0.00/3.30/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 310s ] thds: 256 tps: 1.00 qps: 1.00 (r/w/o: 0.00/1.00/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 320s ] thds: 256 tps: 1.10 qps: 1.10 (r/w/o: 0.00/1.10/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 330s ] thds: 256 tps: 1.10 qps: 1.10 (r/w/o: 0.00/1.10/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 340s ] thds: 256 tps: 1.30 qps: 1.30 (r/w/o: 0.00/1.30/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 350s ] thds: 256 tps: 1.20 qps: 1.20 (r/w/o: 0.00/1.20/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 360s ] thds: 256 tps: 1.30 qps: 1.30 (r/w/o: 0.00/1.30/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 370s ] thds: 256 tps: 1.40 qps: 1.40 (r/w/o: 0.00/1.40/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 380s ] thds: 256 tps: 1.80 qps: 1.80 (r/w/o: 0.00/1.80/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 390s ] thds: 256 tps: 1.90 qps: 1.90 (r/w/o: 0.00/1.90/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 400s ] thds: 256 tps: 2.60 qps: 2.60 (r/w/o: 0.00/2.60/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
[ 410s ] thds: 255 tps: 3.30 qps: 3.30 (r/w/o: 0.00/3.30/0.00) lat (ms,95%): 100000.00 err/s: 0.00 reconn/s: 0.00
SQL statistics:
    queries performed:
        read:                            0
        write:                           984
        other:                           0
        total:                           984
    transactions:                        984    (2.36 per sec.)
    queries:                             984    (2.36 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

General statistics:
    total time:                          417.4665s
    total number of events:              984

Latency (ms):
         min:                                  9.80
         avg:                              99167.12
         max:                             416923.60
         95th percentile:                 100000.00
         sum:                            97580442.66

Threads fairness:
    events (avg/stddev):           3.8438/3.03
    execution time (avg/stddev):   381.1736/34.79
```

When latch scheduler is enabled:

```
sysbench --test=./lua-tests/db/update_non_index.lua --db-driver=mysql --mysql-host=127.0.0.1 --mysql-port=4000 --mysql-user=root --mysql-password= --mysql-db=sbtest1 --oltp-tables-count=1 --oltp-table-size=16 --num-threads=256 --report-interval=10 --max-requests=2000000000 --percentile=95 --max-time=300 run
WARNING: the --test option is deprecated. You can pass a script name or path on the command line without any options.
WARNING: --num-threads is deprecated, use --threads instead
WARNING: --max-requests is deprecated, use --events instead
WARNING: --max-time is deprecated, use --time instead
sysbench 1.0.11 (using system LuaJIT 2.1.0-beta3)

Running the test with following options:
Number of threads: 256
Report intermediate results every 10 second(s)
Initializing random number generator from current time


Initializing worker threads...

Threads started!

[ 10s ] thds: 256 tps: 122.57 qps: 122.57 (r/w/o: 0.00/122.57/0.00) lat (ms,95%): 5607.61 err/s: 0.00 reconn/s: 0.00
[ 20s ] thds: 256 tps: 110.10 qps: 110.10 (r/w/o: 0.00/110.10/0.00) lat (ms,95%): 9624.59 err/s: 0.00 reconn/s: 0.00
[ 30s ] thds: 256 tps: 107.91 qps: 107.91 (r/w/o: 0.00/107.91/0.00) lat (ms,95%): 10722.67 err/s: 0.00 reconn/s: 0.00
[ 40s ] thds: 256 tps: 104.80 qps: 104.80 (r/w/o: 0.00/104.80/0.00) lat (ms,95%): 11115.87 err/s: 0.00 reconn/s: 0.00
[ 50s ] thds: 256 tps: 103.80 qps: 103.80 (r/w/o: 0.00/103.80/0.00) lat (ms,95%): 11523.48 err/s: 0.00 reconn/s: 0.00
[ 60s ] thds: 256 tps: 120.90 qps: 120.90 (r/w/o: 0.00/120.90/0.00) lat (ms,95%): 9977.52 err/s: 0.00 reconn/s: 0.00
[ 70s ] thds: 256 tps: 111.30 qps: 111.30 (r/w/o: 0.00/111.30/0.00) lat (ms,95%): 10917.50 err/s: 0.00 reconn/s: 0.00
[ 80s ] thds: 256 tps: 105.40 qps: 105.40 (r/w/o: 0.00/105.40/0.00) lat (ms,95%): 9799.46 err/s: 0.00 reconn/s: 0.00
[ 90s ] thds: 256 tps: 104.40 qps: 104.40 (r/w/o: 0.00/104.40/0.00) lat (ms,95%): 10722.67 err/s: 0.00 reconn/s: 0.00
[ 100s ] thds: 256 tps: 122.70 qps: 122.70 (r/w/o: 0.00/122.70/0.00) lat (ms,95%): 10722.67 err/s: 0.00 reconn/s: 0.00
[ 110s ] thds: 256 tps: 108.00 qps: 108.00 (r/w/o: 0.00/108.00/0.00) lat (ms,95%): 9977.52 err/s: 0.00 reconn/s: 0.00
[ 120s ] thds: 256 tps: 125.41 qps: 125.41 (r/w/o: 0.00/125.41/0.00) lat (ms,95%): 10531.32 err/s: 0.00 reconn/s: 0.00
[ 130s ] thds: 256 tps: 128.40 qps: 128.40 (r/w/o: 0.00/128.40/0.00) lat (ms,95%): 8955.74 err/s: 0.00 reconn/s: 0.00
[ 140s ] thds: 256 tps: 120.60 qps: 120.60 (r/w/o: 0.00/120.60/0.00) lat (ms,95%): 9624.59 err/s: 0.00 reconn/s: 0.00
[ 150s ] thds: 256 tps: 126.00 qps: 126.00 (r/w/o: 0.00/126.00/0.00) lat (ms,95%): 9452.83 err/s: 0.00 reconn/s: 0.00
[ 160s ] thds: 256 tps: 121.10 qps: 121.10 (r/w/o: 0.00/121.10/0.00) lat (ms,95%): 9624.59 err/s: 0.00 reconn/s: 0.00
[ 170s ] thds: 256 tps: 117.50 qps: 117.50 (r/w/o: 0.00/117.50/0.00) lat (ms,95%): 9799.46 err/s: 0.00 reconn/s: 0.00
[ 180s ] thds: 256 tps: 109.60 qps: 109.60 (r/w/o: 0.00/109.60/0.00) lat (ms,95%): 10531.32 err/s: 0.00 reconn/s: 0.00
[ 190s ] thds: 256 tps: 111.00 qps: 111.00 (r/w/o: 0.00/111.00/0.00) lat (ms,95%): 10531.32 err/s: 0.00 reconn/s: 0.00
[ 200s ] thds: 256 tps: 118.10 qps: 118.10 (r/w/o: 0.00/118.10/0.00) lat (ms,95%): 9799.46 err/s: 0.00 reconn/s: 0.00
[ 210s ] thds: 256 tps: 118.60 qps: 118.60 (r/w/o: 0.00/118.60/0.00) lat (ms,95%): 9284.15 err/s: 0.00 reconn/s: 0.00
[ 220s ] thds: 256 tps: 114.90 qps: 114.90 (r/w/o: 0.00/114.90/0.00) lat (ms,95%): 10722.67 err/s: 0.00 reconn/s: 0.00
[ 230s ] thds: 256 tps: 117.69 qps: 117.69 (r/w/o: 0.00/117.69/0.00) lat (ms,95%): 10722.67 err/s: 0.00 reconn/s: 0.00
[ 240s ] thds: 256 tps: 118.61 qps: 118.61 (r/w/o: 0.00/118.61/0.00) lat (ms,95%): 9799.46 err/s: 0.00 reconn/s: 0.00
[ 250s ] thds: 256 tps: 106.78 qps: 106.78 (r/w/o: 0.00/106.78/0.00) lat (ms,95%): 9284.15 err/s: 0.00 reconn/s: 0.00
[ 260s ] thds: 256 tps: 102.20 qps: 102.20 (r/w/o: 0.00/102.20/0.00) lat (ms,95%): 12384.09 err/s: 0.00 reconn/s: 0.00
[ 270s ] thds: 256 tps: 106.02 qps: 106.02 (r/w/o: 0.00/106.02/0.00) lat (ms,95%): 10722.67 err/s: 0.00 reconn/s: 0.00
[ 280s ] thds: 256 tps: 108.10 qps: 108.10 (r/w/o: 0.00/108.10/0.00) lat (ms,95%): 11115.87 err/s: 0.00 reconn/s: 0.00
[ 290s ] thds: 256 tps: 106.10 qps: 106.10 (r/w/o: 0.00/106.10/0.00) lat (ms,95%): 10917.50 err/s: 0.00 reconn/s: 0.00
[ 300s ] thds: 256 tps: 105.20 qps: 105.20 (r/w/o: 0.00/105.20/0.00) lat (ms,95%): 9799.46 err/s: 0.00 reconn/s: 0.00
SQL statistics:
    queries performed:
        read:                            0
        write:                           34294
        other:                           0
        total:                           34294
    transactions:                        34294  (113.18 per sec.)
    queries:                             34294  (113.18 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

General statistics:
    total time:                          302.9876s
    total number of events:              34294

Latency (ms):
         min:                                  7.52
         avg:                               2249.82
         max:                              47590.55
         95th percentile:                  10343.39
         sum:                            77155206.07

Threads fairness:
    events (avg/stddev):           133.9609/20.42
    execution time (avg/stddev):   301.3875/0.75
```

**There will be near 100x performance improment.**

@disksing

File: store/tikv/latch/latch.go [hunk 5]
```diff
@@ -161,75 +157,140 @@ func (latches *Latches) acquire(lock *Lock) acquireResult {
...
 }
+
+// recycle is not thread safe, the latch should acquire its lock before executing this function.
+func (l *latch) recycle(currentTS uint64) {
+	if l.queue == nil {
+		return
+	}
+
+	prev := l.queue
+	curr := l.queue.next
```
Comment (disksing) [inline store/tikv/latch/latch.go]: Better initialize as `prev = nil, curr = l.queue`, so don't have to handle head node separately. Or you can consider using the indirect pointer hack suggested by Linus Torvalds https://medium.com/@bartobri/applying-the-linus-tarvolds-good-taste-coding-requirement-99749f37684a

Comment (tiancaiamao) [reply store/tikv/latch/latch.go]: Actually, I found that trick myself before I saw any article about it. (I remember I wrote a blog post http://www.zenlife.tk/fake-list-head.md around that time)

**Comment by disksing** [comment]
LGTM.

File: store/tikv/latch/latch.go [hunk 6]
```diff
@@ -161,75 +157,129 @@ func (latches *Latches) acquire(lock *Lock) acquireResult {
...
 }
+
+// recycle is not thread safe, the latch should acquire its lock before executing this function.
+func (l *latch) recycle(currentTS uint64) {
+	fakeHead := node{next: l.queue}
+	prev := &fakeHead
+	for curr := prev.next; curr != nil; curr = curr.next {
+		if tsoSub(currentTS, curr.maxCommitTS) >= expireDuration && curr.value == nil {
+			l.count--
+			prev.next = curr.next
```
Comment (zhangjinpeng1987) [inline store/tikv/latch/latch.go]: How about first node dropped by this function? The latch's queue should point to next node.

File: store/tikv/latch/latch.go [hunk 7]
```diff
@@ -22,32 +23,26 @@ import (
...
-
-func (l *latch) refreshCommitTS(commitTS uint64) {
-	l.Lock()
-	defer l.Unlock()
-	l.maxCommitTS = mathutil.MaxUint64(commitTS, l.maxCommitTS)
+// latch stores a key's waiting transactions information.
+type latch struct {
+	queue   *node
+	count   int
+	waiting []*Lock
```
Comment (zhangjinpeng1987) [inline store/tikv/latch/latch.go]: Why not each node has a waiting queue?

File: store/tikv/latch/latch.go [hunk 8]
```diff
@@ -161,75 +157,129 @@ func (latches *Latches) acquire(lock *Lock) acquireResult {
...
+			break
+		}
 	}
-	return nextLock
-}
+	// Wake up the first one in waiting queue.
+	if idx < len(latch.waiting) {
+		nextLock = latch.waiting[idx]
+		// Delete element latch.waiting[idx] from the array.
+		copy(latch.waiting[idx:], latch.waiting[idx+1:])
```
Comment (zhangjinpeng1987) [inline store/tikv/latch/latch.go]: It's better to use a list for waiting locks.

**Comment by zhangjinpeng1987** [comment]
@tiancaiamao Have you tested sysbench OLTP_RW-pareto scenario?

File: store/tikv/latch/latch.go (hunk 8 above)
Comment (zhangjinpeng1987) [reply store/tikv/latch/latch.go]: @coocood What's your advise?

File: store/tikv/latch/latch.go (hunk 7 above)
Comment (tiancaiamao) [reply store/tikv/latch/latch.go]: Waiting queue is moved from each node to the latch for those reasons:
1. nodes in the queue is inserted every now and then, if each node has is waiting queue, the queue would be created and destroyed. There will be more allocations, and it's less memory efficient.
2. I have an assumption that the waiting queue would not be large, when a list is small enough, an array is very efficient.
3. You may still remember the "first waiting one automatically become running" problem and the old code is complex enough to handle different states. If each node doesn't have the waiting queue, the problem could be avoid. 
@zhangjinpeng1987

**Pull request merged**